datafusion/physical-optimizer/src/enforce_sorting/mod.rs

```diff
@@ -373,9 +373,10 @@ pub fn ensure_sorting(
         return adjust_window_sort_removal(requirements).map(Transformed::yes);
     } else if is_sort_preserving_merge(plan)
         && child_node.plan.output_partitioning().partition_count() <= 1
+        && plan.fetch().is_none()
     {
         // This `SortPreservingMergeExec` is unnecessary, input already has a
-        // single partition.
+        // single partition and no fetch is required.
         let child_node = requirements.children.swap_remove(0);
         return Ok(Transformed::yes(child_node));
```

Comment:
Another possible fix:
```rust
        let mut child_node = requirements.children.swap_remove(0);
        if let Some(fetch) = plan.fetch() {
            // Add the limit exec if the spm has a fetch
            child_node.plan = Arc::new(LocalLimitExec::new(
                Arc::clone(&child_node.plan),
                fetch,
            ));
        }
        return Ok(Transformed::yes(child_node))
```

Reply:
Thanks, this should be more reasonable, so we will only keep the sortexec with fetch after limit pushdown when there is only one partition?

Reply:
ideally, yes